flink-tests/src/test/java/org/apache/flink/test/util/NumberSequenceSourceBlockableByCheckpoint.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.test.util;
+
+import org.apache.flink.api.connector.source.ReaderOutput;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.SplitEnumerator;
+import org.apache.flink.api.connector.source.SplitEnumeratorContext;
+import org.apache.flink.api.connector.source.lib.NumberSequenceSource;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceEnumerator;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceReader;
+import org.apache.flink.api.connector.source.lib.util.IteratorSourceSplit;
+import org.apache.flink.core.io.InputStatus;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+
+/**
+ * A {@link NumberSequenceSource} that can be blocked by checkpoints. If {@link
+ * #isBlockByCheckpoint} is set to true, this source would guarantee that a checkpoint has been
+ * invoked before the source ends its input.
+ */
+public class NumberSequenceSourceBlockableByCheckpoint extends NumberSequenceSource {
```

Comment:
Instead of replacing `TestingNumberSequenceSource` with this new class with the extra ability to set isBlockByCheckpoint to false, it seems simpler to keep `TestingNumberSequenceSource` as is and use `NumberSequenceSource` when needed.

Reply:
I have kept `TestingNumberSequenceSource`'s behavior as it had been. But given that this PR is changing this class from a private class to a public class in test utils package, the name `TestingNumberSequenceSource` might not provide enough information about its function. Thus I named it as `NumberSequenceSourceWithWaitForCheckpoint`.